docs/design-documents/promtail-scraping-objectstore.md

```diff
@@ -0,0 +1,89 @@
+
+## Proposed Solution for the above mentioned points
+
+> There is no appending of lines to file in object store
+
+In the positions.yaml along with the `offset` we should also store file modified `datetime`. If we see that date is modified we should reset the offset and tail the contents from the begining. This is the same behaviour in local filesystem as well. If a file is `opened`, `hpcloud tail` will tail the file from the begining.
+
+#### Alternative 
+
+Instead of tailing the file from the beging upon modification, we can calculate the `checksum` of the contents till the `offset`, and check if the checksum was changed or not after the file was modified. If it is not changed then we can just proceed further, else we can tail from the begining.
+
+Any other better way to handle this?
+
+
+> Where should we keep `positions.yaml` file ?
+
+I think this should be an user choice. One thing we need to consider here is that there will be API limit per account per region and if we store the `positions.yaml` in the object store itself, we may end up doing lot of APIs calls and we would get `RequestLimitExceeded` very frequently. Still, backoff and retry should be implemented in case we get this error even when doing API calls less frequently. So, placing and updating the `positions.yaml` should be left to the user so he can decide based on the API usage of his setup wether to keep it locally or remotely. 
+
+> How we should identify and watch newly added files ?
+
+Limit on API calls should also be considered for syncing the newly added files. To start tailing the newly added files, we should do `ListObjects` at a desired interval. We should make sure we don't keep this interval too small.
+Also, once the entire contents of the file is read we should stop watching the file so that we reduce the number of active `go routines`. 
+
+> How should we support reading of compressed files ?
+
+It would be good to support most of the compression algorithms like we already do for chunks(Snappy, LZ4, Gzip). 
+**Note:** We can only support decompressing of data only if the whole file is compressed using any of the above algorithms, we cannot build support for decompressing data for custom data format stored in files.
+
+## Example S3 config for `positions.yaml`
+
+```yaml
+positions:
+  objectstore:
+    aws:
+      ## bucket name
+      bucket: BUCKET_NAME
+      
+      ## endpoint for the store
+      endpoint: 
+
+      ## secret key
+      secret_access_key: 
+
+      ## access key
+      access_key: 
+
+      ## path to store positions.yaml
+      filename: /path/to/positions.yaml   
+      
```

Comment:
I think we should support a glob to limit path in the bucket for two reasons : 

1)  You can use this for scaling horizontally, aka running multiple promtails each of them with different path. ( That's cool until we have something better, like a ring to shard files across multiple promtails)
2) You may not want to read and send  from all files in the buckets.

Reply:
Agree